metricbeat/module/mongodb/replstatus/replstatus_integration_test.go

```diff
@@ -0,0 +1,140 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build integration
+
+package replstatus
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/tests/compose"
+	mbtest "github.com/elastic/beats/metricbeat/mb/testing"
+	"github.com/elastic/beats/metricbeat/module/mongodb"
+)
+
+func TestFetch(t *testing.T) {
+	compose.EnsureUp(t, "mongodb")
+
+	err := initiateReplicaSet(t)
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	f := mbtest.NewEventFetcher(t, getConfig())
+	event, err := f.Fetch()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+
+	t.Logf("%s/%s event: %+v", f.Module().Name(), f.Name(), event)
+
+	// Check event fields
+	oplog := event["oplog"].(common.MapStr)
+	allocated := oplog["size"].(common.MapStr)["allocated"].(int64)
+	assert.True(t, allocated >= 0)
+
+	used := oplog["size"].(common.MapStr)["used"].(float64)
+	assert.True(t, used > 0)
+
+	firstTs := oplog["first"].(common.MapStr)["timestamp"].(int64)
+	assert.True(t, firstTs >= 0)
+
+	window := oplog["window"].(int64)
+	assert.True(t, window >= 0)
+
+	members := event["members"].(common.MapStr)
+	primary := members["primary"].(common.MapStr)
+	assert.NotEmpty(t, primary["host"].(string))
+	assert.True(t, primary["optime"].(int64) > 0)
+
+	set := event["set_name"].(string)
+	assert.Equal(t, set, "beats")
+}
+
+func TestData(t *testing.T) {
+	compose.EnsureUp(t, "mongodb")
+
+	f := mbtest.NewEventFetcher(t, getConfig())
+	err := mbtest.WriteEvent(f, t)
+	if err != nil {
+		t.Fatal("write", err)
+	}
+}
+
+func getConfig() map[string]interface{} {
+	return map[string]interface{}{
+		"module":     "mongodb",
+		"metricsets": []string{"replstatus"},
+		"hosts":      []string{mongodb.GetEnvHost() + ":" + mongodb.GetEnvPort()},
+	}
+}
+
+func initiateReplicaSet(t *testing.T) error {
+	url := getConfig()["hosts"].([]string)[0]
+
+	dialInfo, err := mgo.ParseURL(url)
+	if err != nil {
+		return err
+	}
+	dialInfo.Direct = true
+
+	mongoSession, err := mongodb.NewDirectSession(dialInfo)
+	if err != nil {
+		return err
+	}
+	defer mongoSession.Close()
+
+	// get oplog.rs collection
+	db := mongoSession.DB("admin")
+	config := ReplicaConfig{"beats", []Host{{0, url}}}
+	var initiateResult map[string]interface{}
+	if err := db.Run(bson.M{"replSetInitiate": config}, &initiateResult); err != nil {
+		if err.Error() != "already initialized" {
+			return err
+		}
+	}
+
+	var status map[string]interface{}
+	for {
+		db.Run(bson.M{"replSetGetStatus": 1}, &status)
+		myState, ok := status["myState"].(int)
+		t.Logf("Mongodb state is %d", myState)
+		if ok && myState == 1 {
+			time.Sleep(5 * time.Second) // hack, wait more for replica set to become stable
+			break
+		}
```

Comment:
Could we add a sleep after every retry?

Reply:
Yes we can, but I see no reason to do this.
As you mentioned before, we should wait for some condition instead of sleeping. First I thought we should wait for a primary node, and I expected this condition to be sufficient for running the test. But then -in action- I figured out it needs something more than a node in the primary state. As I didn't find any condition to wait for, I used this sleep expresion.

If you're worried about blowing up CPU with this loop, I should note that state changes very fast (like 5 3 2 2 1) and it doesn't seem to be an issue.

Reply:
Ok, let's leave it like this by now.